lib/Service/RecentPagesService.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+declare(strict_types=1);
+
+namespace OCA\Collectives\Service;
+
+use OCA\Collectives\Model\PageInfo;
+use OCA\Collectives\Model\RecentPage;
+use OCP\DB\Exception;
+use OCP\DB\QueryBuilder\IQueryBuilder;
+use OCP\Files\IMimeTypeLoader;
+use OCP\IConfig;
+use OCP\IDBConnection;
+use OCP\IURLGenerator;
+use OCP\IUser;
+
+class RecentPagesService {
+
+	public function __construct(
+		protected CollectiveService $collectiveService,
+		protected IDBConnection $dbc,
+		protected IConfig $config,
+		protected IMimeTypeLoader $mimeTypeLoader,
+		protected IURLGenerator $urlGenerator,
+	) {	}
+
+	/**
+	 * @return RecentPage[]
+	 * @throws MissingDependencyException
+	 * @throws Exception
+	 */
+	public function forUser(IUser $user, int $limit = 10): array {
+		try {
+			$collectives = $this->collectiveService->getCollectives($user->getUID());
+		} catch (NotFoundException|NotPermittedException) {
+			return [];
+		}
+
+		$qb = $this->dbc->getQueryBuilder();
+		$appData = $this->getAppDataFolderName();
+		$mimeTypeMd = $this->mimeTypeLoader->getId('text/markdown');
+
+		$expressions = [];
+		foreach ($collectives as $collective) {
+			$value = sprintf($appData . '/collectives/%d/%%', $collective->getId());
+			$expressions[] = $qb->expr()->like('f.path', $qb->createNamedParameter($value, IQueryBuilder::PARAM_STR));
+		}
+
+		$qb->select('p.*', 'f.mtime as timestamp', 'f.name as filename', 'f.path as path')
+			->from('filecache', 'f')
+			->leftJoin('f', 'collectives_pages', 'p', $qb->expr()->eq('f.fileid', 'p.file_id'))
+			->where($qb->expr()->orX(...$expressions))
+			->andWhere($qb->expr()->eq('f.mimetype', $qb->createNamedParameter($mimeTypeMd, IQueryBuilder::PARAM_INT)))
+			->orderBy('f.mtime', 'DESC')
+			->setMaxResults($limit);
+
+		$r = $qb->executeQuery();
+
+		$pages = [];
+		$collectives = [];
+		while ($row = $r->fetch()) {
+			$collectiveId = (int)explode('/', $row['path'], 4)[2];
+			if (!isset($collectives[$collectiveId])) {
+				try {
+					// collectives are not cached, but always read from DB, so keep them
+					$collectives[$collectiveId] = $this->collectiveService->getCollectiveInfo($collectiveId, $user->getUID());
```

Comment:
If `$collectives` from above wouldn't be overwritten, we could reuse the `CollectiveInfo` object from that array and wouldn't have to do an extra `getCollectiveInfo()` here, no?